snakemake/scheduler.py

```diff
@@ -355,18 +355,17 @@ def stats(self):
         except AttributeError:
             raise TypeError("Executor does not support stats")
 
-    def candidate(self, job):
-        """ Return whether a job is a candidate to be executed. """
-        return (
-            job not in self.running
-            and job not in self.failed
-            and (self.dryrun or (not job.dynamic_input and not self.dag.dynamic(job)))
-        )
-
     @property
     def open_jobs(self):
         """ Return open jobs. """
-        return filter(self.candidate, list(job for job in self.dag.ready_jobs))
+        jobs = self.dag.ready_jobs - (self.running | self.failed)
```

Comment:
Are you sure this is faster? I would've thought that the additional allocations instead make it slower. Did you benchmark that?

Reply:
Yes, the set operations are written in C,  making them much faster then a custom loop. I  just did some benchmarking, which shows that the set-based approach is ~10 times as fast (set sizes do not have a large impact on this ratio). 

But you are correct that it does some additional allocations. I now rewrote this to avoid the (running | failed) allocation, which provides another 33% gain in speed.